File: Source code/app/lib/models/trace_event.dart
class TraceEvent {
  final String id;
  final String product;
  final String batchId;
  final String eventType;
  final String description;
  final Map<String, dynamic>? details;
  final List<String> images;
  final List<String> videos;
  final String? dataHash;
  final String? txHash;
  final int? blockNumber;
  final String onChainStatus; // pending | confirmed | failed
  final DateTime createdAt;
  final Map<String, dynamic>? recordedBy;

  TraceEvent({
    required this.id,
    required this.product,
    required this.batchId,
    required this.eventType,
    required this.description,
    this.details,
    required this.images,
    this.videos = const [],
    this.dataHash,
    this.txHash,
    this.blockNumber,
    required this.onChainStatus,
    required this.createdAt,
    this.recordedBy,
  });

  factory TraceEvent.fromJson(Map<String, dynamic> json) => TraceEvent(
    id: json['_id'] as String,
    product: json['product'] is Map
        ? json['product']['_id'] as String
        : json['product'] as String,
    batchId: json['batchId'] as String? ?? '',
    eventType: json['eventType'] as String,
    description: json['description'] as String? ?? '',
    details: json['details'] as Map<String, dynamic>?,
    images: (json['images'] as List<dynamic>? ?? const [])
        .map(
          (item) => item is Map<String, dynamic>
              ? (item['path'] ?? item['url'] ?? '').toString()
              : item.toString(),
        )
        .where((item) => item.isNotEmpty)
        .toList(),
    videos: (json['videos'] as List<dynamic>? ?? const [])
        .map(
          (item) => item is Map<String, dynamic>
              ? (item['path'] ?? item['url'] ?? '').toString()
              : item.toString(),
        )
        .where((item) => item.isNotEmpty)
        .toList(),
    dataHash: json['dataHash'] as String?,
    txHash: json['txHash'] as String?,
    blockNumber: json['blockNumber'] as int?,
    onChainStatus: json['onChainStatus'] as String? ?? 'pending',
    createdAt: DateTime.parse(
      json['createdAt'] as String? ?? DateTime.now().toIso8601String(),
    ),
    recordedBy: json['recorded_by'] as Map<String, dynamic>?,
  );

  String get eventLabel => _eventLabels[eventType] ?? eventType;

  String get eventIcon => _eventIcons[eventType] ?? '📋';

  /// Tên người ghi nhận
  String get recordedByName {
    if (recordedBy == null) return '';
    final first = recordedBy!['first_name'] ?? '';
    final last = recordedBy!['last_name'] ?? '';
    return '$first $last'.trim();
  }

  static const _eventLabels = {
    'SEEDING': 'Gieo hạt',
    'FERTILIZING': 'Bón phân',
    'WATERING': 'Tưới nước',
    'PEST_CONTROL': 'Kiểm soát sâu bệnh',
    'PESTICIDE': 'Phun thuốc BVTV',
    'HARVESTING': 'Thu hoạch',
    'PROCESSING': 'Chế biến',
    'PACKAGING': 'Đóng gói',
    'SHIPPING': 'Vận chuyển',
    'QUALITY_CHECK': 'Kiểm tra chất lượng',
  };

  static const _eventIcons = {
    'SEEDING': '🌱',
    'FERTILIZING': '🧪',
    'WATERING': '💧',
    'PEST_CONTROL': '🐛',
    'PESTICIDE': '💊',
    'HARVESTING': '🌾',
    'PROCESSING': '🏭',
    'PACKAGING': '📦',
    'SHIPPING': '🚚',
    'QUALITY_CHECK': '✅',
  };
}

File: Source code/api/src/models/TraceEvent.ts
import { Schema, model, Types } from 'mongoose';

export const ACTION_TYPES = [
  'SEEDING',
  'FERTILIZING',
  'WATERING',
  'PEST_CONTROL',
  'HARVESTING',
  'PACKAGING',
  'SHIPPING',
] as const;

export type ActionType = (typeof ACTION_TYPES)[number];

export interface ITraceEvent {
  product: Types.ObjectId;
  batchId: string;
  eventType: ActionType;
  description: string;
  details?: Record<string, unknown>;
  images: { path: string; filename: string }[];
  videos: { path: string; filename: string; mimeType?: string }[];
  recorded_by: Types.ObjectId;
  dataHash?: string;
  dataHashVersion?: 'v1' | 'v2';
  txHash?: string;
  blockNumber?: number;
  onChainStatus: 'pending' | 'confirmed' | 'failed' | 'skipped';
  actionIndex?: number;
  createdAt: Date;
  updatedAt: Date;
}

const traceEventSchema = new Schema<ITraceEvent>(
  {
    product: {
      type: Schema.Types.ObjectId,
      ref: 'Product',
      required: true,
    },
    batchId: {
      type: String,
      required: true,
      index: true,
    },
    eventType: {
      type: String,
      required: [true, 'Vui l?ng ch?n lo?i h?nh ??ng'],
      enum: ACTION_TYPES,
    },
    description: {
      type: String,
      required: [true, 'Vui l?ng nh?p m? t?'],
    },
    details: {
      type: Schema.Types.Mixed,
      default: {},
    },
    images: {
      type: [{ path: String, filename: String }],
      default: [],
    },
    videos: {
      type: [{ path: String, filename: String, mimeType: String }],
      default: [],
    },
    recorded_by: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    dataHash: { type: String },
    dataHashVersion: {
      type: String,
      enum: ['v1', 'v2'],
    },
    txHash: { type: String },
    blockNumber: { type: Number },
    onChainStatus: {
      type: String,
      enum: ['pending', 'confirmed', 'failed', 'skipped'],
      default: 'pending',
    },
    actionIndex: { type: Number },
  },
  { timestamps: true }
);

traceEventSchema.index({ product: 1, createdAt: 1 });

export default model<ITraceEvent>('TraceEvent', traceEventSchema);
